File: dart/test/parameter_type_registry_test.dart
import 'package:cucumber_expressions/src/errors.dart';
import 'package:cucumber_expressions/src/parameter_type.dart';
import 'package:cucumber_expressions/src/parameter_type_registry.dart';
import 'package:test/test.dart';

class Name {
  Name(this.name);
  final String name;
}

class Person {
  Person(this.name);
  final String name;
}

class Place {
  Place(this.name);
  final String name;
}

final RegExp capitalisedWord = RegExp(r'[A-Z]+\w+');

void main() {
  group('ParameterTypeRegistry', () {
    late ParameterTypeRegistry registry;

    setUp(() {
      registry = ParameterTypeRegistry();
    });

    test('does not allow more than one preferential parameter type per regexp',
        () {
      registry
        ..defineParameterType(
          ParameterType<Name>(
            'name',
            capitalisedWord,
            'Name',
            (s) => Name(s.first!),
            useForSnippets: true,
            preferForRegexpMatch: true,
          ),
        )
        ..defineParameterType(
          ParameterType<Person>(
            'person',
            capitalisedWord,
            'Person',
            (s) => Person(s.first!),
            useForSnippets: true,
            preferForRegexpMatch: false,
          ),
        );
      expect(
        () => registry.defineParameterType(
          ParameterType<Place>(
            'place',
            capitalisedWord,
            'Place',
            (s) => Place(s.first!),
            useForSnippets: true,
            preferForRegexpMatch: true,
          ),
        ),
        throwsA(
          isA<CucumberExpressionException>().having(
            (e) => e.message,
            'message',
            equals(
              'There can only be one preferential parameter type per regexp. '
              'The regexp /${capitalisedWord.pattern}/ is used for two '
              'preferential parameter types, {name} and {place}',
            ),
          ),
        ),
      );
    });

    test('looks up preferential parameter type by regexp', () {
      final name = ParameterType<Name>(
        'name',
        RegExp(r'[A-Z]+\w+'),
        'Name',
        (s) => Name(s.first!),
        useForSnippets: true,
        preferForRegexpMatch: false,
      );
      final person = ParameterType<Person>(
        'person',
        RegExp(r'[A-Z]+\w+'),
        'Person',
        (s) => Person(s.first!),
        useForSnippets: true,
        preferForRegexpMatch: true,
      );
      final place = ParameterType<Place>(
        'place',
        RegExp(r'[A-Z]+\w+'),
        'Place',
        (s) => Place(s.first!),
        useForSnippets: true,
        preferForRegexpMatch: false,
      );

      registry
        ..defineParameterType(name)
        ..defineParameterType(person)
        ..defineParameterType(place);

      expect(
        registry.lookupByRegexp(
          r'[A-Z]+\w+',
          RegExp(r'([A-Z]+\w+) and ([A-Z]+\w+)'),
          'Lisa and Bob',
        ),
        same(person),
      );
    });
  });
}

File: elixir/test/cucumber/cucumber_expressions/parameter_type_registry_test.exs
defmodule Cucumber.CucumberExpressions.ParameterTypeRegistryTest do
  use ExUnit.Case, async: true

  alias Cucumber.CucumberExpressions.{ParameterType, ParameterTypeRegistry}

  @capitalised_word "[A-Z]+\\w+"

  defp type(name, prefer) do
    ParameterType.new!(
      name: name,
      regexps: @capitalised_word,
      type: :string,
      prefer_for_regexp_match: prefer
    )
  end

  test "does not allow more than one preferential parameter type per regexp" do
    registry =
      ParameterTypeRegistry.new()
      |> ParameterTypeRegistry.define_parameter_type!(type("name", true))
      |> ParameterTypeRegistry.define_parameter_type!(type("person", false))

    assert {:error, error} =
             ParameterTypeRegistry.define_parameter_type(registry, type("place", true))

    assert Exception.message(error) ==
             "There can only be one preferential parameter type per regexp. " <>
               "The regexp /[A-Z]+\\w+/ is used for two: {name} and {place}"
  end

  test "looks up preferential parameter type by regexp" do
    registry =
      ParameterTypeRegistry.new()
      |> ParameterTypeRegistry.define_parameter_type!(type("name", false))
      |> ParameterTypeRegistry.define_parameter_type!(type("person", true))
      |> ParameterTypeRegistry.define_parameter_type!(type("place", false))

    looked_up =
      ParameterTypeRegistry.lookup_by_regexp(
        registry,
        @capitalised_word,
        "([A-Z]+\\w+) and ([A-Z]+\\w+)",
        "Lisa and Bob"
      )

    assert looked_up.name == "person"
  end

  test "raises ambiguous exception when no parameter types are preferential" do
    registry =
      ParameterTypeRegistry.new()
      |> ParameterTypeRegistry.define_parameter_type!(type("name", false))
      |> ParameterTypeRegistry.define_parameter_type!(type("person", false))
      |> ParameterTypeRegistry.define_parameter_type!(type("place", false))

    error =
      assert_raise Cucumber.CucumberExpressions.AmbiguousParameterTypeError, fn ->
        ParameterTypeRegistry.lookup_by_regexp(
          registry,
          @capitalised_word,
          "([A-Z]+\\w+) and ([A-Z]+\\w+)",
          "Lisa and Bob"
        )
      end

    assert Exception.message(error) ==
             "Your Regular Expression /([A-Z]+\\w+) and ([A-Z]+\\w+)/\n" <>
               "matches multiple parameter types with regexp /[A-Z]+\\w+/:\n" <>
               "   {name}\n   {person}\n   {place}\n" <>
               "\n" <>
               "I couldn't decide which one to use. You have two options:\n" <>
               "\n" <>
               "1) Use a Cucumber Expression instead of a Regular Expression. Try one of these:\n" <>
               "   {name} and {name}\n" <>
               "   {name} and {person}\n" <>
               "   {name} and {place}\n" <>
               "   {person} and {name}\n" <>
               "   {person} and {person}\n" <>
               "   {person} and {place}\n" <>
               "   {place} and {name}\n" <>
               "   {place} and {person}\n" <>
               "   {place} and {place}\n" <>
               "\n" <>
               "2) Make one of the parameter types preferential and continue to use a Regular Expression.\n" <>
               "\n"
  end
end
